Eeprom data

Tiogaplanet opened the pull request:
Got setUserData() and getUserData() working, provided two example sketches and updated documentation.

Commit:
Merge remote-tracking branch 'refs/remotes/adamgreen/master'

Commit:
Merge remote-tracking branch 'refs/remotes/adamgreen/master' into EepromData

Commit:
Trying to get read/write EEPROM working

Wrote setUserData() and gerUserData().

Commit:
Incremental improvement to reading/writing EEPROM

I found a couple obvious bugs in creating the mip command array to send
to the mip.  Added some documentaton.  Will link example in readme
later.

Commit:
Enabled setUserData() and getUserData()

Implemented the setUserData and getUserData functions, provided example
sketches and updated documentation.

adamgreen commented on Arduino/MiP_ProMini_Pack_Library/src/mip.cpp:
```diff
@@ -1814,7 +1816,90 @@ int8_t MiP::rawGetGameMode(MiPGameMode& mode)
+
+
+
+void MiP::setUserData(uint8_t address, uint8_t userData)
```
Might be easier for the user if you switched the first parameter from being an address to being an offset and only accept values between 0x00 and 0x0F. You could then add 0x20 before calling down to the raw APIs. You can use MIP_ASSERT() to verify that they keep the offset between 0x00 an 0x0F.

Since you are only calling your verifiedSetUserData() function from one place, you can probably just pull that code up into this function directly and get rid of verifiedSetUserData().

adamgreen commented on Arduino/MiP_ProMini_Pack_Library/src/mip.cpp:
```diff
@@ -1814,7 +1816,90 @@ int8_t MiP::rawGetGameMode(MiPGameMode& mode)
+    verifiedSetUserData(address, userData);
+}
+
+int8_t MiP::getUserData(uint8_t address, uint8_t& userData)
```
If you use existing APIs like readRadar(), readGestureEvent(), readVolume(), etc which return simple integers or enumerations as an example then it would probably be better to get rid of the userData parameter and just return the read value instead. I don't return errors from most of the public functions since the average Arduino user is just going to assume that their calls succeeded and won't write error handling code. That is why I added all of the read after write and retries down in the library so that I can isolate the average Arduino developer from as much error handling as possible.

I would continue to have rawGetUserData() to use the userData parameter and return an error code since that is consistent with most of the existing raw APIs.

adamgreen commented on Arduino/MiP_ProMini_Pack_Library/src/mip.cpp:
```diff
@@ -1814,7 +1816,90 @@ int8_t MiP::rawGetGameMode(MiPGameMode& mode)
+    {
+        return result;
+    }
+    if (responseLength != 3 || response[0] != MIP_CMD_GET_USER_DATA)
```
Can you be extra careful here and verify that ```response[1]``` equals ```address``` as well?

adamgreen reviewed (changes requested):
Cool to see that you got this working! Awesome!

Tiogaplanet commented:
OK, I've address all of your review comments and amended my commit.  Added defines for the EEPROM address, used offsets for public functions and updated documentation accordingly.

adamgreen reviewed (approved):
Looks great!
